### PR TITLE
Allocate players into teams on the basis of cumulative team scores

### LINKED ISSUE #315: Take player stats into consideration during team allocation
Player's are evenly allocated into their teams solely on the basis of player count, without giving any weightage to their stats whatsoever. This results in a haphazard distribution of skill levels across the two teams.

Please also discuss the criteria to be taken into consideration, and other implementation details.

### PR DESCRIPTION
**Pull request by ClobberXD**
- `ctf.autoalloc` mode is set to 4 in `minetest.conf`. (MT-CTF/ctf_pvp_engine#30)
- `ctf.custom_alloc` function is overridden in `ctf_alloc/init.lua`.
- Cumulative team scores are tracked and updated on each player join and leave.
- Team with lowest cumulative score is stored separately, and is updated if required, when a player joins or leaves the team.
- `ctf.custom_alloc` just returns the name of the team with the lowest cumulative score (which has already been calculated).
- `ctf_alloc.set_all` (for mass alloc of all online players at start of new match) is only modified to call `calc_scores()`, which updates the cumulative scores of all teams.

I've also added a bunch of print statements to help with debugging, which I'll remove/convert to proper minetest.log calls before merging.

I've tested this PR (final test results [here](https://github.com/MT-CTF/capturetheflag/pull/319#issuecomment-541403715)) and it works, but it could do with more testing. Closes #315.

### CODE REVIEW

**Comment by ClobberXD** [comment]
Upon testing, I noticed that:

- There are no teams available
- There's no match-related mechanisms like build timer
- Player spawns at (0,0,0) and isn't moved to team spawn (expected, due to no.1)
- No schems are placed

I'm sure I'm missing something very trivial, which will fix all of this. WIP for now.

**Comment by ClobberXD** [comment]
Aha, I see the problem here. `ctf_match` and `ctf_map` aren't loaded due to unsatisfied dependencies.

- `ctf_alloc` depends on `ctf_stats`
- `ctf_match` depends on `ctf_alloc`
- `ctf_stats` depends on `ctf_match`

**Comment by ClobberXD** [comment]
Removing `ctf_stats` from `depends.txt` fixes this issue (as expected), although I'd prefer a better solution.

Otherwise, this PR works. I can only test with a max. of 2 players due to the limitations of my potato PC, but the player who joins second is allocated to the empty team.

**Comment by ClobberXD** [comment]
Forgot to push the latest changes. Can be tested now.

**Comment by ClobberXD** [comment]
Rebased

**Comment by MinetestSam** [comment]
Tested it but the cumulative scores of both the teams were very different
screenshot_20190708_192325
Note: BOT1 and BOT2 ended up in the same team whereas BOT and MTSam were in the other team.
The correct teams according to cumulative scores should have been BOT and BOT1 in one team and BOT2 and MTSam in the other.

**Comment by ClobberXD** [comment]
Pushed the latest changes (had forgotten to do it, again). I've also optimized the hell out of this PR, and simplified the custom alloc function. I'll update the first post to reflect the current state of changes.

**Comment by ClobberXD** [comment]
PR seems to be working as expected.

### Testing procedure

- Tested with 3 players:
  - ANAND (score: 436433)
  - ClobberXD (score: 352.5)
  - RedTomato (score: 0)
- ANAND joins first; gets allocated to team blue. 
  ```
  red: 0
  blue: 436433
  ```
- ClobberXD joins next; gets allocated to team red. Red is the team with lowest cumulative score.
  ```
  red: 352.5
  blue: 436433
  ```
- RedTomato joins next; gets allocated to team red, as that's the team with lowest cumulative team score.
  ```
  red: 352.5
  blue: 436433
  ```

### Console output

```
[CaptureTheFlag] (autoalloc) ANAND was allocated to blue
Cumulative scores of teams: {
	blue = 436433
}
Setting lowest to [blue,436433]
[calc_scores] Computing cumulative scores for all teams
	red:
	Cumulative team score: 0
	********
[update_lowest] Setting lowest to [red,0]
	blue:
		436433 (ANAND)
	Cumulative team score: 436433
	********
[calc_scores] ********
2019-07-09 09:57:45: ACTION[Server]: ClobberXD joins game. List of players: ANAND ClobberXD
[CaptureTheFlag] (autoalloc) ClobberXD was allocated to red
Cumulative scores of teams: {
	red = 352.5,
	blue = 436433
}
[CaptureTheFlag] (autoalloc) RedTomato was allocated to red
Cumulative scores of teams: {
	red = 352.5,
	blue = 436433
}
2019-07-09 10:40:18: ACTION[Server]: RedTomato joins game. List of players: ClobberXD ANAND RedTomato
Cumulative scores of teams: {
	red = 352.5,
	blue = 436433
}
[CaptureTheFlag] (team) Removing player "RedTomato"
2019-07-09 10:42:21: ACTION[Server]: RedTomato leaves game. List of players: ANAND ClobberXD

(RedTomato was kicked due to inactivity, lol)

2019-07-09 10:46:38: ACTION[Server]: RedTomato joins game. List of players: ANAND ClobberXD RedTomato
[CaptureTheFlag] (autoalloc) RedTomato was allocated to red
Cumulative scores of teams: {
	red = 352.5,
	blue = 436433
}
```

**Comment by MinetestSam** [comment]
Tested again and it works. Console Output https://pastebin.com/WFxWMJYA

**Comment by ClobberXD** [comment]
MT-CTF/ctf_pvp_engine#40 should simplify this PR even further. This PR shouldn't be merged until that one is merged.

**Comment by ClobberXD** [comment]
***BUMP***

**Comment by ClobberXD** [comment]
I don't want to wait for MT-CTF/ctf_pvp_engine#40, because:

- This PR has already waited long enough.
- The aforementioned PR has been closed, and I don't have the time to re-implement it, because I recently discovered that the original `on_leave_team` implementation was incomplete, and implementing it properly would require a lot more code to be tweaked.

****

@MT-CTF/developers This PR is ready for review. Note that this PR has been tested many times already, by people other than me. Code-wise, it can't get simpler than this. If you have time to review PRs, please review this one first.

**Comment by Thomas--S** [comment]
I think I found a bug:

1) Player A has a low score and joins/starts the game first.
2) A is allocated to red.
3) Player B has a high score and joins second.
4) B is allocated to blue.
5) Player C has now score and joins third.
6) C is allocated to blue.


Log output:

```
 - creating red
[CaptureTheFlag] (team) Creating team red
[CaptureTheFlag] (flag) Adding flag to red at (8373, 1, 96)
 - creating blue
[CaptureTheFlag] (team) Creating team blue
[CaptureTheFlag] (flag) Adding flag to blue at (8439, -1, -96)
2019-10-03 15:51:58: WARNING[Server]: {
	{
		player = <userdata>,
		score = 2535
	}
}
[CaptureTheFlag] (autoalloc) Thomas was allocated to red
2019-10-03 15:51:58: ACTION[Server]: Thomas joined team red
Cumulative scores of teams: {
	red = 2535
}
Setting lowest to [red,2535]
[CaptureTheFlag] (flag) red has wrong node at flag position, ignore, correcting...
[CaptureTheFlag] (flag) blue has wrong node at flag position, ignore, correcting...

[calc_scores] Computing cumulative scores for all teams
	red:
		2535 (Thomas)
	Cumulative team score: 2535
	********
	blue:
	Cumulative team score: 0
	********
Setting lowest to [blue,0]
[calc_scores] ********

2019-10-03 15:51:58: ACTION[Server]: Giving initial stuff to player Thomas
2019-10-03 15:52:02: WARNING[Emerge-0]: Placing 70 chests from (8301,-25,-99) to (8499,30,-1)
2019-10-03 15:52:02: WARNING[Emerge-0]: Spawned 69/70 chests after 82 attempts!
2019-10-03 15:52:02: WARNING[Emerge-0]: Placing 70 chests from (8500,-25,100) to (8300,30,0)
2019-10-03 15:52:02: WARNING[Emerge-0]: Spawned 70/70 chests after 73 attempts!
[CaptureTheFlag] (flag) red has wrong node at flag position, air, correcting...
[CaptureTheFlag] (flag) blue has wrong node at flag position, air, correcting...
2019-10-03 15:52:07: ACTION[Server]: Thomas-S [127.0.0.1] joins game. 
2019-10-03 15:52:07: ACTION[Server]: Thomas-S joins game. List of players: Thomas Thomas-S
[CaptureTheFlag] (autoalloc) Thomas-S was allocated to blue
2019-10-03 15:52:07: ACTION[Server]: Thomas-S joined team blue
Cumulative scores of teams: {
	red = 2535,
	blue = 52340
}
2019-10-03 15:52:07: ACTION[Server]: Giving initial stuff to player Thomas-S
{
	r = 255,
	b = 255,
	a = 0,
	g = 255
}
[CaptureTheFlag] (autoalloc) testing was allocated to blue
2019-10-03 15:54:04: ACTION[Server]: testing joined team blue
Cumulative scores of teams: {
	red = 2535,
	blue = 52340
}
2019-10-03 15:54:04: ACTION[Server]: testing [127.0.0.1] joins game. 
2019-10-03 15:54:04: ACTION[Server]: testing joins game. List of players: Thomas Thomas-S testing
2019-10-03 15:54:04: ACTION[Server]: Giving initial stuff to player testing
{
	r = 255,
	b = 255,
	a = 0,
	g = 255
}
Gauge not found!
Cumulative scores of teams: {
	red = 2535,
	blue = 52340
}
[CaptureTheFlag] (team) Removing player "testing"
2019-10-03 15:56:28: ACTION[Server]: testing leaves game. List of players: Thomas Thomas-S 
```

As far as I can see, the problem is in the `update_lowest(team)` function.

In `ctf.register_on_join_team()` this function is called with `team` being the team to which the new player got allocated, i.e. the lowest team.
However, the `update_lowest(team)` only can do two things:
1) Don't change the lowest team.
2) Set the lowest team to `team`, but as we learned above, `team` is already the lowest team most of the times.

**Comment by ClobberXD** [comment]
Fixed the bug, and considerably simplified the PR, at the cost of a negligible performance impact - cumulative team scores are computed only when `ctf.custom_alloc` is invoked.

I've also removed the unused lock-to-team parts.

****

<details>
<summary>Test 1 (lower score player joins first)</summary>

Tested with the following join order:

- ClobberXD (592.5)
- ANAND (~436763)
- RedTomato (0)

Test results:

```
[CaptureTheFlag] (team) Creating team red
[CaptureTheFlag] (team) Creating team blue

[CaptureTheFlag] (autoalloc) ClobberXD was allocated to red

[calc_scores]
	red = 592.5
	blue = 0

lowest = {
	team = "blue",
	score = 0
}

[calc_scores] ********

2019-10-13 14:44:31: ACTION[Server]: ANAND joins game. List of players: ClobberXD ANAND

[calc_scores]
	red = 592.5
	blue = 0

lowest = {
	team = "blue",
	score = 0
}

[calc_scores] ********

[CaptureTheFlag] (autoalloc) ANAND was allocated to blue

2019-10-13 14:44:37: ACTION[Server]: RedTomato joins game. List of players: ANAND ClobberXD RedTomato

[calc_scores]
	red = 592.5
	blue = 436763

lowest = {
	team = "red",
	score = 592.5
}

[calc_scores] ********

[CaptureTheFlag] (autoalloc) RedTomato was allocated to red
```

</details>

<details>
<summary>Test 2 (higher score player joins first)</summary>

Tested with the following join order:

- ANAND (~436763)
- ClobberXD (592.5)
- RedTomato (0)

Test results:

```
[CaptureTheFlag] (team) Creating team red
[CaptureTheFlag] (team) Creating team blue

[CaptureTheFlag] (autoalloc) ANAND was allocated to red

[calc_scores]
	red = 436763
	blue = 0

lowest = {
	team = "blue",
	score = 0
}

[calc_scores] ********

2019-10-13 15:15:10: ACTION[Server]: ClobberXD joins game. List of players: ANAND ClobberXD

[calc_scores]
	red = 436763
	blue = 0

lowest = {
	team = "blue",
	score = 0
}

[calc_scores] ********

[CaptureTheFlag] (autoalloc) ClobberXD was allocated to blue

2019-10-13 15:15:15: ACTION[Server]: RedTomato joins game. List of players: ClobberXD ANAND RedTomato

[calc_scores]
	red = 436763
	blue = 592.5

lowest = {
	team = "blue",
	score = 592.5
}

[calc_scores] ********

[CaptureTheFlag] (autoalloc) RedTomato was allocated to blue
```

</details>

**Comment by Thomas--S** [comment]
Probably alloc mode 4 should be documented here: https://github.com/MT-CTF/capturetheflag/blob/master/docs/doc_settings.md

---

Do I understand it correctly, that the all-players-allocation at the beginning of every match is still done by just adding one player to red, the next one to blue etc. (or vice versa)?

If so, may I ask: Why?

Wouldn't it be better to just call `ctf.autoalloc()` for every player?

---

I haven't tested this yet, but the code looks much better now.

**Comment by ClobberXD** [comment]
> Probably alloc mode 4 should be documented here: https://github.com/MT-CTF/capturetheflag/blob/master/docs/doc_settings.md

Probably...

> Do I understand it correctly, that the all-players-allocation at the beginning of every match is still done by just adding one player to red, the next one to blue etc. (or vice versa)?
> 
> If so, may I ask: Why?
> 
> Wouldn't it be better to just call `ctf.autoalloc()` for every player?

This PR, comprising of roughly 80 changed lines (a huge proportion of them isn't even code), is already 9 months and 13 days old. That's why. ;)

To be clear, I'm not blaming anything or anyone for the slow pace of development - that's how things are, and no one can take the blame for that. But it's quite obvious that the more complex a PR is, merging it takes exponentially more time. That's why I try to keep my PRs as small and as concise as possible. This particular PR already requires more testing than usual. Using the new allocation technique for mass allocation would require even more testing, and this PR would be merged only around the time when my grand-children get through high-school. :D

File: mods/ctf/ctf_alloc/init.lua [hunk 1]
```diff
@@ -81,4 +100,6 @@ function ctf_alloc.set_all()
 
 		player:set_hp(20)
 	end
+
+	calc_scores()
```
Comment (Thomas--S) [inline mods/ctf/ctf_alloc/init.lua]: I just wonder why this is needed :thinking: 

This was also why I came to the idea of questioning this function.

Comment (ClobberXD) [reply mods/ctf/ctf_alloc/init.lua]: Good question... I suppose it just exists as a relic of the past. I'll remove it - it shouldn't affect anything, because `calc_scores` is still called in `ctf.custom_alloc` before returning `lowest.team`.

Comment (ClobberXD) [reply mods/ctf/ctf_alloc/init.lua]: Done

**Comment by ClobberXD** [comment]
Re-added manual team-locking but moved the code to a separate file, because why not?

**Comment by Thomas--S** [comment]
To be honest, I liked the old solution of overriding the allocation function better because it worked independently of the allocation mode.

Additionally, I'd prefer it if such refactoring was done in a follow-up PR because a small diff makes reviewing easier.

**Comment by ClobberXD** [comment]
I'd do anything (short of closing this PR) to make reviewing easier for you, sire. o7

File: mods/ctf/ctf_alloc/init.lua [hunk 2]
```diff
@@ -1,16 +1,17 @@
 local storage = minetest.get_mod_storage()
 local data = minetest.parse_json(storage:get_string("locktoteam")) or {}
 
-local ctf_autoalloc = ctf.autoalloc
-function ctf.autoalloc(name, alloc_mode)
+-- Override autoalloc function to implement team-locking
+ctf_autoalloc = ctf.autoalloc
```
Comment (Thomas--S) [inline mods/ctf/ctf_alloc/init.lua]:
```suggestion
local ctf_autoalloc = ctf.autoalloc
```

Comment (ClobberXD) [reply mods/ctf/ctf_alloc/init.lua]: I wasn't done yet, but thanks, lol

**Comment by ClobberXD** [comment]
@Thomas--S I've moved the team-locking code back into `init.lua`. PR should be ready for review now.

File: mods/ctf/ctf_alloc/init.lua [hunk 3]
```diff
@@ -29,6 +30,63 @@ end, {
...
+	end
+
+	update_lowest()
+	print("[calc_scores] ********\n")
+end
+
+-- Override team-allocation logic
+-- Allocate player into the team with the lowest cumulative score
+function ctf.custom_alloc(name)
+	local team = ctf_alloc.get_locked_team(name)
```
Comment (Thomas--S) [inline mods/ctf/ctf_alloc/init.lua]: I think this line is part of an unrelated change that is already reverted. I wasn't able to find such a function anywhere in the code.

Comment (ClobberXD) [reply mods/ctf/ctf_alloc/init.lua]: Correct, my bad.

Comment (ClobberXD) [reply mods/ctf/ctf_alloc/init.lua]: Reverted, thanks

File: mods/ctf/ctf_alloc/init.lua [hunk 4]
```diff
@@ -1,16 +1,17 @@
 local storage = minetest.get_mod_storage()
 local data = minetest.parse_json(storage:get_string("locktoteam")) or {}
 
+-- Override autoalloc function to implement team-locking
 local ctf_autoalloc = ctf.autoalloc
-function ctf.autoalloc(name, alloc_mode)
+function ctf.autoalloc(name)
```
Comment (Thomas--S) [inline mods/ctf/ctf_alloc/init.lua]: May I ask why the second parameter was removed here but not from the original function?

This seems rather inconsistent to me.

Comment (ClobberXD) [reply mods/ctf/ctf_alloc/init.lua]: Erm, you were saying...? :D
